include/oneapi/dpl/internal/histogram_impl.h

```diff
@@ -0,0 +1,117 @@
+                                                                __histogram_first, __num_bins, __func,
+                                                                ::std::forward<_Range>(__opt_range)...);
+    }
+}
+#endif // _ONEDPL_BACKEND_SYCL
+
+template <typename _ExecutionPolicy, typename _RandomAccessIterator1, typename _RandomAccessIterator2, typename _Size,
+          typename _IdxHashFunc, typename... _Range>
+inline oneapi::dpl::__internal::__enable_if_host_execution_policy<typename ::std::decay<_ExecutionPolicy>::type>
+__pattern_histogram(_ExecutionPolicy&& exec, _RandomAccessIterator1 __first, _RandomAccessIterator1 __last,
+                    _RandomAccessIterator2 __histogram_first, const _Size& __num_bins, _IdxHashFunc __func,
+                    _Range&&... __opt_range)
+{
+    assert(false && "Histogram API is not currently supported for host parallel policies");
+}
+
+template <typename _ExecutionPolicy, typename _RandomAccessIterator1, typename _RandomAccessIterator2, typename _Size,
+          typename _T>
+inline void
+__histogram_impl(_ExecutionPolicy&& exec, _RandomAccessIterator1 __first, _RandomAccessIterator1 __last,
+                 _RandomAccessIterator2 __histogram_first, const _Size& __num_bins, const _T& __first_bin_min_val,
+                 const _T& __last_bin_max_val)
+{
+    __internal::__pattern_histogram(
+        ::std::forward<_ExecutionPolicy>(exec), __first, __last, __histogram_first, __num_bins,
+        __internal::__evenly_divided_binhash<_T>(__first_bin_min_val, __last_bin_max_val, __num_bins));
+}
+
+template <typename _ExecutionPolicy, typename _RandomAccessIterator1, typename _RandomAccessIterator2, typename Iter3>
+inline void
+__histogram_impl(_ExecutionPolicy&& exec, _RandomAccessIterator1 __first, _RandomAccessIterator1 __last,
+                 _RandomAccessIterator2 __histogram_first, Iter3 __boundary_first, Iter3 __boundary_last)
+{
+    auto boundary_view = oneapi::dpl::__ranges::guard_view<Iter3>(__boundary_first, __boundary_last);
+    __internal::__pattern_histogram(::std::forward<_ExecutionPolicy>(exec), __first, __last, __histogram_first,
+                                    boundary_view.size() - 1, __internal::__custom_range_binhash{boundary_view});
+}
+
+} // namespace __internal
+
+template <typename _ExecutionPolicy, typename _RandomAccessIterator1, typename _Size, typename _T,
+          typename _RandomAccessIterator2>
+oneapi::dpl::__internal::__enable_if_execution_policy<_ExecutionPolicy, _RandomAccessIterator2>
+histogram(_ExecutionPolicy&& exec, _RandomAccessIterator1 __first, _RandomAccessIterator1 __last,
+          const _Size& __num_bins, const _T& __first_bin_min_val, const _T& __last_bin_max_val,
+          _RandomAccessIterator2 __histogram_first)
+{
+    __internal::__histogram_impl(::std::forward<_ExecutionPolicy>(exec), __first, __last, __histogram_first, __num_bins,
```

Comment:
The reason is not satisfying.  It is because the public API changed later in the process to place the output at the end, and the details did not.  

I've fixed it so the details match the public interface up until converting to ranges, thanks.